relay-general/src/store/normalize/measurements.rs

```diff
@@ -0,0 +1,147 @@
+        }
+    }
+}
+
+type OperationName = String;
+
+type OperationNameIntervals = HashMap<OperationName, Vec<TimeWindowSpan>>;
+
+fn merge_intervals(intervals: &Vec<TimeWindowSpan>) -> Vec<TimeWindowSpan> {
+    let mut intervals = intervals.clone();
+
+    // sort by start_timestamp in ascending order
+    intervals.sort_unstable_by(|a, b| a.start_timestamp.partial_cmp(&b.start_timestamp).unwrap());
+
+    intervals.into_iter().fold(
+        vec![],
+        |mut merged, current_interval| -> Vec<TimeWindowSpan> {
+            // merged is a vector of disjoint intervals
+
+            if merged.len() == 0 {
+                merged.push(current_interval);
+                return merged;
+            }
+
+            let mut last_interval = merged.last_mut().unwrap();
+
+            if last_interval.end_timestamp < current_interval.start_timestamp {
+                // if current_interval does not overlap with last_interval,
+                // then add current_interval
+                merged.push(current_interval);
+                return merged;
+            }
+
+            // current_interval and last_interval overlaps; so we merge these intervals
+
+            // invariant: last_interval.start_timestamp <= current_interval.start_timestamp
+
+            last_interval.end_timestamp =
+                std::cmp::max(last_interval.end_timestamp, current_interval.end_timestamp);
+
+            merged
+        },
+    )
+}
+
+/// Ensure measurements interface is only present for transaction events and emit operation breakdown measurements
+pub fn normalize_measurements(event: &mut Event) {
+    if event.ty.value() != Some(&EventType::Transaction) {
+        // Only transaction events may have a measurements interface
+        event.measurements = Annotated::empty();
+        return;
+    }
+
+    // Generate operation breakdowns
+    if let Some(spans) = event.spans.value() {
+        let intervals: OperationNameIntervals =
+            spans
+                .iter()
+                .fold(HashMap::new(), |mut intervals, span| match span.value() {
+                    None => return intervals,
+                    Some(span) => {
+                        let cover = TimeWindowSpan::new(
+                            span.start_timestamp.value().unwrap().clone(),
+                            span.timestamp.value().unwrap().clone(),
+                        );
+
+                        let operation_name: OperationName = span.op.value().unwrap().clone();
+
+                        if !intervals.contains_key(&operation_name) {
+                            intervals.insert(operation_name, vec![cover]);
+                            return intervals;
+                        }
+
+                        if let Some(operation_name_interval) = intervals.get_mut(&operation_name) {
+                            operation_name_interval.push(cover);
+                            *operation_name_interval = merge_intervals(operation_name_interval);
```

Comment:
My rust eyes aren't that great since it's been a while, but are we doing merge_intervals for every span we're iterating over? If so, should we just merge intervals once after all spans have been pushed to the map?